### ISSUE TITLE
trtexec very slow when running on GPU Tesla T4

### ISSUE THREAD

**Issue by dmortem** [NVIDIA/TensorRT#3075]
## Description

I tried to convert my onnx model to tensorRT using trtexec, but it takes me a long time (around 12.3h for a model with size of 24M for tensorrt and 14M for onnx). Any suggestion on this? Thank you!

For example, can I skip the profiling step to save time? 

## Environment

**TensorRT Version**: 8.6.1.6-1+cuda12.0

**NVIDIA GPU**: Tesla T4

**NVIDIA Driver Version**: 530.30.02

**CUDA Version**: cuda_12.1.r12.1/compiler.32688072_0

**CUDNN Version**: 8.9.1

Operating System:

Python Version (if applicable): 3.8.12

PyTorch Version (if applicable): 1.13.1

## Relevant Files

<img width="1649" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/20540613/f397b98c-0a84-4f17-9040-c731c94383d9">

**Model link**:

**Comment by zerollzeng** [comment]
Looks like this is not a small model, it has *65988* Node(s)
```
$ polygraphy inspect model trtexec_slow.onnx
[I] Loading model: /home/scratch.zeroz_sw/github_bug/3075/trtexec_slow.onnx
[I] ==== ONNX Model ====
    Name: torch_jit | ONNX Opset: 14

    ---- 2 Graph Input(s) ----
    {view_1 [dtype=float32, shape=(5, 2048, 33)],
     view_2 [dtype=float32, shape=(5, 2048, 33)]}

    ---- 1 Graph Output(s) ----
    {signal [dtype=float32, shape=(5,)]}

    ---- 42 Initializer(s) ----

    ---- 65988 Node(s) ----

```

**Comment by dmortem** [comment]
Thanks for your reply! The model is written by myself and I convert pytorch model to onnx by calling `torch_onnx.export(model, (batch, {}), "model.onnx", verbose=False, input_names=input_names, output_names=output_names, export_params=True)`. 

So your point is since there are too many onnx operators, the conversion from onnx to tensorrt is quite slow? My pytorch model is not large, only 269866 parameters by calling `sum(p.numel() for p in model.parameters())`. In fact, I have some torch operators like unfold in my model. I am not sure whether these operators lead to huge number of operators in onnx.

Therefore, if I want to speed up the conversion from pytorch model to tensorrt, do you have any suggestions?

**Comment by dmortem** [comment]
By the way, may I ask how you applied constant folding? I noticed that the default value of the argument `do_constant_folding` is True for `torch_onnx.export`.

**Comment by zerollzeng** [comment]
Yeah, Here is how TRT build an engine: first onnx parser parse the onnx to TRT INetworkDefinition, then TRT will do graph optimization(layer fusion, dead node elimination, etc.) then for the optimized graph, TRT test performance for all available tactics for each fused kernel. then find the best tactic and combination globally.

To speed up building time, I think you can refer to https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#opt-builder-optimization-level

**Comment by zerollzeng** [comment]
https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#timing-cache should also be helpful in your case.

**Comment by dmortem** [comment]
Hi, thanks for your reply.

I tried several things:
- polygraphy surgeon sanitize onnx with constant folding first, and then convert onnx to tensorrt. But the building time is even a little slower. (46602 seconds vs 44435 seconds)
<img width="1656" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/20540613/61089164-629c-4270-b6d6-fad409e913b1">

- set optimization level to 0, `trtexec --onnx=model.onnx --saveEngine=model.trt --builderOptimizationLevel=0 --verbose`, building time is only a little faster.
- use time cache. I found trtexec "default is to enable timing cache", I also tried `trtexec --onnx=model.onnx --timingCacheFile=timing_cache.txt --saveEngine=model.trt ----verbose` to serialize the cache, but it doesn't work when I convert onnx to tensorrt using the existing time cache.

I put some logs of trtexec https://drive.google.com/drive/folders/1Fw8jFtcSirlgBu7wrksvVA-PQ7Lsh2rp?usp=sharing. 
- "trt_normal.txt" is the original conversion log, normal model size, without constant folding
- "trt_normal_fold.txt" is the conversion log of the onnx model with constant folding
- "trt_tiny.txt" is the conversion log of the tiny model 
- "trt_tiny_optim0.txt" is the conversion log of the tiny model with building flag builderOptimizationLevel=0
- "trt_tiny_optim1_gen_timingcache.txt" is the conversion log of the tiny model with building flag builderOptimizationLevel=1 and exporting timing cache file
- "trt_tiny_optim2_use_timeingcache.txt" is the conversion log of the tiny model with building flag builderOptimizationLevel=2 and using saved timing cache file

I think currently the main bottleneck is the profiling time of `ForeignNode[onnx::Gather_82...(Unnamed Layer* 1510) [ElementWise]]`, but the log of this part is limited and I don't know how to decrease this profiling time.

In fact, I wonder whether there is a way to only export the weights of a pytorch model to an existing tensorrt. Specifically, if the network structure of the pytorch model keeps unchanged, and the weights will be updated every week, can I spend 100 hours to convert pytorch model to tensorrt for the first time, and then simply export the weights to the existing tensorrt engine every week after the weights of pytorch model are updated?

What is more, I found trtexec can only utilize one cpu, is it possible for trtexec to use more cpus to speed up the building time?

I appreciate any help you can provide, thanks!

**Comment by zerollzeng** [comment]
@nvpohanh @zhenhuaw-me Do you have any suggestions? Thanks! ^ ^

**Comment by zerollzeng** [comment]
> In fact, I wonder whether there is a way to only export the weights of a pytorch model to an existing tensorrt. Specifically, if the network structure of the pytorch model keeps unchanged, and the weights will be updated every week, can I spend 100 hours to convert pytorch model to tensorrt for the first time, and then simply export the weights to the existing tensorrt engine every week after the weights of pytorch model are updated?

Yes, we support refit. see https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#refitting-engine-c

**Comment by dmortem** [comment]
Refit seems a good choice! I will try it, thanks!

**Comment by dmortem** [comment]
@nvpohanh @zhenhuaw-me 
Hi do you have any suggestions about my slow conversion problem? Thanks!

**Comment by zhenhuaw-me** [comment]
11h sounds a bit too long... Which GPU clock do you set and what's CPU are you using?

**Comment by dmortem** [comment]
GPU clock: 
<img width="773" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/20540613/fad15589-617f-44a4-969c-a809aa2137cc">

CPU: 
<img width="1909" alt="image" src="https://github.com/NVIDIA/TensorRT/assets/20540613/67bf8b87-9c2a-40ef-bd2a-bddf7f9eb26e">

In fact, this model is only a submodule of my real model... So if there is a way to save time, it will be really helpful for me :)

**Comment by zhenhuaw-me** [comment]
Does the refit feature work for you?
If it's possible to share your model thus we can further analyze?

**Comment by Feynman1999** [comment]
any update on this issue? i meet the same question

**Comment by zhenhuaw-me** [comment]
Please try these two solutions:
1. Use TRT 9.2 release. Find download link here: https://github.com/NVIDIA/TensorRT/issues/3604#issuecomment-1900052173
2. You may try [`nvinfer1::IBuilderConfig::setBuilderOptimizationLevel()`](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_builder_config.html#af51d2c96a3dcc98c345141ff93562e42) to speed up your engine building. To be noted, the engine inference time may increase.
